files/verify-cn.py

```diff
@@ -19,94 +20,152 @@
+    """Determine if a user should be allowed to connect."""
+    # Create the FreeIPA client
+    ipa_client: ClientMeta = ClientMeta(dns_discovery=realm)
+
+    # Login client in using kerberos credentials
+    logging.debug(f"Logging with kerberos to IPA server for realm: {realm}")
+    ipa_client.login_kerberos()
+
+    logging.debug("Searching for user with matching certificate.")
+    response = ipa_client.certmap_match(client_certificate)
+    logging.debug(f"Received response from FreeIPA: {response}")
+
+    if response["count"] == 0:
+        logging.warning("No matching user found.")
         return False
 
-    if len(rec) > 1:
-        logging.warning(f"More than 1 record found in LDAP.  Count: {len(rec)}.")
+    if response["count"] > 1:
+        logging.warning(
+            f"More than 1 user matched certificate.  Count: {response['count']}."
+        )
         return False
 
-    user_dn, attribute_list = rec[0]
-    logging.info(f"Found user in LDAP: {user_dn}")
-    users_groups = attribute_list.get("memberOf")
-    # Check to see if user is a member of the correct group
-    permit_user = vpngroup_dn.encode("utf-8") in users_groups
-    if permit_user:
-        logging.info(f"User is member of {vpngroup_dn}")
+    # Extract username from response
+    matched_uid: str = response["result"][0]["uid"][0]
+    logging.info(f"Certificate matched uid: {matched_uid}")
+
+    # Get user data from FreeIPA
+    logging.debug(f"Looking up user record for: {matched_uid}")
+    user = ipa_client.user_find(matched_uid)["result"][0]
+    logging.debug(f"User record: {user}")
+
+    group_ok: bool
+    account_enabled: bool
+    account_not_preserved: bool
+
+    # Check to see if user is a member of the group
+    if group in user["memberof_group"]:
+        logging.debug(f"{matched_uid} is member of {group}")
+        group_ok = True
+    else:
+        logging.warning(f"{matched_uid} is NOT a member of {group}")
+        group_ok = False
+
+    # Check to see if the user's account is active
+    if user["nsaccountlock"] is False:
+        logging.debug(f"{matched_uid} account is not locked.")
+        account_enabled = True
     else:
-        logging.warning(f"User IS NOT member of {vpngroup_dn}")
-    return permit_user
+        logging.warning(f"{matched_uid} account IS LOCKED.")
+        account_enabled = False
 
+    # Check to see if the user's account is active
+    if user["preserved"] is False:
+        logging.debug(f"{matched_uid} account is not preserved.")
+        account_not_preserved = True
+    else:
+        logging.warning(f"{matched_uid} account IS PRESERVED.")
+        account_not_preserved = False
 
-def main():
+    # Pass judgement on the user
+    if group_ok and account_enabled and account_not_preserved:
+        logging.info(f"{matched_uid} will be permitted access.")
+        return True
+    else:
+        logging.warning(f"{matched_uid} ACCESS DENIED.")
```

Comment:
⛔⛔⛔ 👍

Reply:
Previously @mcdonnnj pointed out that there are good reasons to prefer format strings to f-strings in `logging` calls like this.  Not sure if you had forgotten this or were purposefully disregarding it.  We should as a team choose one style and use it everywhere.

Reply:
Forgot about that- we should avoid f-strings in logging states.

https://blog.pilosus.org/posts/2020/01/24/python-f-strings-in-logging/

Reply:
Begrudgingly adjusted by @felddy in https://github.com/cisagov/ansible-role-openvpn/pull/18/commits/3b0a4f629408e4c26f5227eee74117e2705f3415. 😜

Reply:
😏